### DISCUSSION TITLE
DLL TwinBasic

### DISCUSSION THREAD

**Discussion by siggismalz** [joyfullservice/msaccess-vcs-addin#530] [General]
Hello, can anyone explain how TwinBasic works?

**Comment by joyfullservice** [comment]
[TwinBASIC](https://twinbasic.com/) is essentially a new development IDE/compiler for writing VBA/BASIC type code and compiling to EXE/DLL. I use it in this project because it provides a very simple and maintainable way to compile a helper COM add-in for the VBA IDE that acts as a wrapper to launch the Microsoft Access add-in through a ribbon interface. The other key advantage is that TwinBASIC allows us to create a 64-bit add-in that works in 64-bit Microsoft Access.

In this project, the code in the TwinBASIC COM add-in is pretty minimal, keeping as much as possible in the native Access VBA code in the Access add-in. The VCS ribbon can be extended and customized through the XML file without having to recompile the COM add-in, which is a big plus especially as we look towards supporting multiple languages in the future.

Hope that helps!

**Reply by siggismalz** [reply]
Can I extend the ribbon with functionalities without touching the DLL?
I can't find the references to the onaction button in xml.

**Reply by joyfullservice** [reply]
Yes, absolutely! That is the advantage of the approach we took with it. The `onAction` events all link to the same generic `OnActionButton` function.

![image](https://github.com/user-attachments/assets/da3be277-90b9-4a46-af9d-d7ce61480472)

When you click the button, the _button name_ is passed over to the add-in, the prefix stripped off, and it looks for a matching function in `clsVersionControl`. For example, the button above calls the following code:

```vba
'---------------------------------------------------------------------------------------
' Procedure : ExportSelected
' Author    : Adam Waller
' Date      : 2/21/2023
' Purpose   : Export the selected object
'---------------------------------------------------------------------------------------
'
Public Sub ExportSelected()
    
    Dim objSelected As AccessObject
    
    If HasFormOpen Then Exit Sub
    
    Set objSelected = GetSelectedNavPaneObject
    If objSelected Is Nothing Then
        MsgBox2 "Please Select an Object First", _
            "Select a single object in the Navigation Pane to export.", _
            "(This item must have the keyboard focus.)", vbInformation
    Else
        ' Export the item
        RunExport , objSelected
    End If

End Sub
```

To add new buttons to the ribbon, use a unique name for the button, and create a corresponding sub in the `clsVersionControl` class.
